jupyterhub/values.yaml

```diff
@@ -331,18 +331,11 @@ singleuser:
   networkPolicy:
     enabled: true
     ingress: []
-    egress:
-      # Required egress to communicate with the hub and DNS servers will be
-      # augmented to these egress rules.
-      #
-      # This default rule explicitly allows all outbound traffic from singleuser
-      # pods, except to a typical IP used to return metadata that can be used by
-      # someone with malicious intent.
-      - to:
-          - ipBlock:
-              cidr: 0.0.0.0/0
-              except:
-                - 169.254.169.254/32
+    egressAllowRules:
+      cloudMetadataServer: false
+      dnsPorts: true
+      publicIPs: true
```

Comment:
Maybe `onlyPublicIPs` or similar, to indicate that we are in fact blocking private IPs here?

Reply:
Oh nooo, I really really dislike the idea of communicating that we are blocking private IPs, we are not blocking them, we are allowing public IPs!

It is such a common confusion with network policies, when enabled to target a pod with regards to egress, they are isolated and nothing is allowed. That is the only operation that restricts the pods networking, after that, it is all about allowing. This allows publicIPs, it doesn't say "only" public IPs, it just allows them. If we have another rule like "dnsPorts" we allow communication to dnsPorts as well, even they are not public IPs for example.

Do you see my point? I'm very stubborn about avoiding the use of "block", "disallow", "only", "restrict" etc, because we are only declaring what we allow here, not what we disallow.

Reply:
Yeah, communicating around network policies is tricky and I appreciate the thought you are putting into naming things! 

The breaking change in this PR - and hence z2jh 2.x vs 1.x - is that private IPs will not be accessible anymore. Public IP behavior has not changed, while private IP behavior changes. I want the naming to communicate this - hence my suggestion of `onlyPublicIP`.

The change in this PR has two parts:

1. Allow access to 0.0.0.0/0 (which is basically *everything*)
2. But restrict access to three private IP ranges

So from a deployer perspective, the change from before is that now we have restricted access to three private IP ranges, while public IP behavior has not changed.

I'd love to hear a bit more about this statement:

>  I'm very stubborn about avoiding the use of "block", "disallow", "only", "restrict" etc, because in this specific configuration, we are only declaring what we allow here, not what we disallow.

Particularly in https://github.com/jupyterhub/zero-to-jupyterhub-k8s/pull/2508/files#diff-70983994b57e310d348c747400dd1ae5ea9f3e4efe63da0155789c3a6bd2a411R132, we do explicitly disallow private IPs. We even use the word 'block' to communicate that in the comment, because that's what it is doing. So understanding a bit more about your mental model there will help me.

Regardless, this is a complex to communicate change because network policies can be a bit complex sometimes, so I appreciate the thoughtfullness here.

Reply:
Thanks for iterating with me @yuvipanda, in the typical very thoughtful and kind way!

Note that the PRs current state does not reflect the communication I intend it to include, so there is a lot of room for communication to be added in the changelog and in general about how network policy configuration can be understood and made.

I'm thinking a communication about this should exist in multiple places:
- changelog
- config reference:
  - networkpolicy level
    - egressAllowRules level
      - specific egressAllowRules level
- documentation: about network config

> The breaking change in this PR - and hence z2jh 2.x vs 1.x - is that private IPs will not be accessible anymore. Public IP behavior has not changed, while private IP behavior changes. I want the naming to communicate this - hence my suggestion of onlyPublicIP.

We are in full agreement about communicating this properly, the question is how. NetworkPolicies are like we agree very complicated to comprehend and work with, and due to that, I care about this detail a lot as well. I want all rules under the egressAllowRules to clearly convey its about ADDING permissions by steering very clear from anything indicating its about restricting permissions - we are in practice just adding the permissions of the public ip range, even though it is done technically by saying "all ips: except the private ips". To help mitigate your concern, i want to add documentation and clarifications about things as I already intended in various places, but I don't want to risk miscommunicating that this may block private IPs when it is in practice just only allowing the public IPs.

If you had a rule that said `onlyPublicIP` it indicates to me that we _disallow_ non-public IPs, while in practice, we are simply just not explicitly allowing them - there is no disallow operation in a NetworkPolicy and this is what I'm super afraid of miscommunicating about. Example to make a point related to this:

```yaml
    # Rule 1
    - to:
        - ipBlock:
            cidr: 10.0.0.0/8
            # The "except" clause within this rule, is just to
            # restrict the range of what we are allowing in
            # this specific rule, out of many rules that add
            # more and more permissions. No rule is
            # removing or restricting permissions!
            except:
              - 10.128.0.0/10
    # Rule 2: equivalent to rule 1, both allow the ranges
    # 10.0.0.0-10.127.255.255 and 10.192.0.0-10.255.255.255
    - to:
        - ipBlock:
            cidr: 10.0.0.0/10
        - ipBlock:
            cidr: 10.64.0.0/10
        - ipBlock:
            cidr: 10.192.0.0/10
```

I also end up thinking that if we name it `onlyPublicIPs`, we could just as well also name `dnsPorts` as `onlyDnsPorts` as well, because i see no technical difference between the situations. Both _add_ permissions about dnsPorts and publicIPs specifically, only about that, but none are disallowing or blocking anything.

Reply:
I think my confusion is around this step:

> I don't want to risk miscommunicating that this may block private IPs when it is in practice just only allowing the public IPs.

But we are explicitly disallowing private IPs. NetworkPolicies *do* have the ability to *disallow* access via the `except` block, which is what we are using. 

Perhaps the problem is that we are trying to name *one* config item to really mean two different things - allowing all IPs (the 0.0.0.0/0) and blocking all private IPs.  Perhaps we can split that into two configurations? Although I'm not sure what an allow rule for `0.0.0.0/0` *without* the `except` block will approach.

> I also end up thinking that if we name it onlyPublicIPs, we could just as well also name dnsPorts as onlyDnsPorts as well, because i see no technical difference between the situations. Both add permissions about dnsPorts and publicIPs specifically, only about that, but none are disallowing or blocking anything.

Ah, now I understand your concern with the addition of the word `only`. I am confused about the publicIP rule not blocking anything though, as we have an `except` clause there.

Reply:
From a deployer perspective, maybe looking through a couple config options might help.

For example, let's look at this config:

```yaml
egressAllowRules:
   publicIPs: false
```

With this config, as an admin, I expect the user pods to just not be able to communicate to any public IPs. However, in practice, I think this will just allow all outgoing traffic, to private and public IPs.

```yaml
egressAllowRules:
   publicIPs: true
```

With this, I expect the change to be that public IPs are accessible. While that is true, private IPs are now blocked - and that isn't apparent from the config.

Perhaps an option is:

```yaml
egressAllowRules:
   publicIPs: true
   privateIPs: false
```

We turn this into two config options, rather than one. We default to `publicIPs: true` and `privateIPs: false`, so same effect as now. Here is a matrix of how that might work

|  | privateIPs: true | privateIPs: false |
| - | - | - |
| publicIPs: true | No network policy setup |   Network Policy that allows *only* public IP ranges (defined as 0.0.0.0/0 - private ranges) (present behavior, default) |
| publicIPs: false | Network Policy that allows *only* private IP ranges | NetworkPolicy that prevents *all* outbound network access (how to do this?) |

Might be a bit more work, but clearer?

Reply:
> But we are explicitly disallowing private IPs. NetworkPolicies do have the ability to disallow access via the except block, which is what we are using.

No i disagree, they can't disallow, they can just make a rule that excludes allowing something while allowing a larger set. That rule is a pure allow-rule.

> Although I'm not sure what an allow rule for 0.0.0.0/0 without the except block will approach.

Yepp, this is nonsensical. This is because it is not a disallow rule, it is a "but as part of this allow rule, don't also allow this section...".

> Ah, now I understand your concern with the addition of the word only. I am confused about the publicIP rule not blocking anything though, as we have an except clause there.

YES! This has contributed a lot to my confusion in the past, and I'm very concerned about adding to this confusion. 

> Network policies do not conflict; they are additive. If any policy or policies apply to a given pod for a given direction, the connections allowed in that direction from that pod is the union of what the applicable policies allow.

This quote was from https://kubernetes.io/docs/concepts/services-networking/network-policies/. It's all about adding permissions that add up in a "union" like operation and no subtraction operation is involved. The `except` clause is just to restrict the specific rule because its very convenient to allow a large set of IP ranges except a small subset rather than building up the range specification piece by piece. For example, if you were to specify all IPs except a single one, you need a lot of IP ranges specified - but if you were instead to say "0.0.0.0/0 except 123.123.123.123/32" it is far easier.

```yaml
    # Rule 1, a rule that allows all but 123.123.123.123/32
    # but it doesn't disallow123.123.123.123/32
    - to:
        - ipBlock:
            cidr: 0.0.0.0/0
            except:
              - 123.123.123.123/32
    # Rule 2: equivalent to rule 1, both allow the same ranges
    - to:
        - ipblock:
            cidr: 0.0.0.0/2
        - ipblock:
            cidr: 64.0.0.0/3
        - ipblock:
            cidr: 96.0.0.0/4
        - ipblock:
            cidr: 112.0.0.0/5
        - ipblock:
            cidr: 120.0.0.0/7
        - ipblock:
            cidr: 122.0.0.0/8
        - ipblock:
            cidr: 123.0.0.0/10
        - ipblock:
            cidr: 123.64.0.0/11
        - ipblock:
            cidr: 123.96.0.0/12
        - ipblock:
            cidr: 123.112.0.0/13
        - ipblock:
            cidr: 123.120.0.0/15
        - ipblock:
            cidr: 123.122.0.0/16
        - ipblock:
            cidr: 123.123.0.0/18
        - ipblock:
            cidr: 123.123.64.0/19
        - ipblock:
            cidr: 123.123.96.0/20
        - ipblock:
            cidr: 123.123.112.0/21
        - ipblock:
            cidr: 123.123.120.0/23
        - ipblock:
            cidr: 123.123.122.0/24
        - ipblock:
            cidr: 123.123.123.0/26
        - ipblock:
            cidr: 123.123.123.64/27
        - ipblock:
            cidr: 123.123.123.96/28
        - ipblock:
            cidr: 123.123.123.112/29
        - ipblock:
            cidr: 123.123.123.120/31
        - ipblock:
            cidr: 123.123.123.122/32
        - ipblock:
            cidr: 123.123.123.124/30
        - ipblock:
            cidr: 123.123.123.128/25
        - ipblock:
            cidr: 123.123.124.0/22
        - ipblock:
            cidr: 123.123.128.0/17
        - ipblock:
            cidr: 123.124.0.0/14
        - ipblock:
            cidr: 123.128.0.0/9
        - ipblock:
            cidr: 124.0.0.0/6
        - ipblock:
            cidr: 128.0.0.0/1
```

Reply:
> For example, let's look at this config:
> 
> ```yaml
> egressAllowRules:
>    publicIPs: false
> ```
> 
> With this config, as an admin, I expect the user pods to just not be able to communicate to any public IPs. However, in practice, I think this will just allow all outgoing traffic, to private and public IPs.

Hmmm... no. We set `egress: []` as a default in this PR, which stops having a default rule that no longer adds the allow rule "0.0.0.0/0 except the cloud metadata server ip". In other words, with that set to false, there is no rule left adding permissions to public or private IPs, but there is a rule adding permissions to dns ports and important pods like the hub.

Reply:
I think the resolution of this discussion is https://github.com/jupyterhub/zero-to-jupyterhub-k8s/pull/2508#issuecomment-998753082.